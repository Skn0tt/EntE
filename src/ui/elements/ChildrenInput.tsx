/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import { Grid, Button } from "@material-ui/core";
import IconButton from "@material-ui/core/IconButton/IconButton";
import AddIcon from "@material-ui/icons/Add";
import DeleteIcon from "@material-ui/icons/Delete";
import List from "@material-ui/core/List/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import ListItemSecondaryAction from "@material-ui/core/ListItemSecondaryAction";
import Typography from "@material-ui/core/Typography/Typography";
import * as _ from "lodash";
import { SearchableDropdown } from "../components/SearchableDropdown";
import { UserN } from "../redux";
import { useState } from "react";
import { makeTranslationHook } from "../helpers/makeTranslationHook";

export const translation = {
  en: {
    title: "Children",
    addChildren: "Add children",
    child: "Child",
  },
  de: {
    title: "Kinder",
    addChildren: "Kinder hinzufÃ¼gen",
    child: "Kind",
  },
};

const useTranslation = makeTranslationHook(translation);

export const includes = (exclude: UserN[]) => {
  const excludedIds = exclude.map((u) => u.get("id"));
  return (u: UserN) => _.includes(excludedIds, u.get("id"));
};

function getInitiallySelectedStudent(props: ChildrenInputProps) {
  return _.differenceBy(props.students, props.children, (u) => u.get("id"))[0];
}

interface ChildrenInputProps {
  children: UserN[];
  students: UserN[];
  onChange: (children: UserN[]) => void;
  text?: {
    title: string;
    addChildren: string;
    child: string;
  };
}

function ChildrenInput(props: ChildrenInputProps) {
  const { children, text, students } = props;
  const [selected, setSelected] = useState<UserN | undefined>(
    getInitiallySelectedStudent(props)
  );

  const [input, setInput] = useState("");

  const translation = useTranslation();

  return (
    <Grid container direction="column">
      <Grid item>
        <Typography variant="h6">{text?.title ?? translation.title}</Typography>
      </Grid>
      {/* List Children */}
      <Grid item>
        <List>
          {children.map((user) => (
            <ListItem key={user.get("id")}>
              <ListItemText primary={user.get("displayname")} />
              <ListItemSecondaryAction>
                <IconButton
                  onClick={() => props.onChange(_.without(children, user))}
                >
                  <DeleteIcon />
                </IconButton>
              </ListItemSecondaryAction>
            </ListItem>
          ))}
        </List>
      </Grid>
      {/* Add Children */}
      <Grid item container>
        <Grid item xs={11}>
          <SearchableDropdown<UserN>
            value={input}
            onChange={setInput}
            items={students.filter((u) => !includes(children)(u))}
            onSelect={setSelected}
            itemToString={(i) => i.get("displayname")}
            includeItem={(u, searchTerm) =>
              u
                .get("displayname")
                .toLowerCase()
                .includes(searchTerm.toLowerCase())
            }
            helperText={!!text ? text.addChildren : translation.addChildren}
            label={!!text ? text.child : translation.child}
          />
        </Grid>
        <Grid item xs={1}>
          <Button
            variant="fab"
            mini
            onClick={() => {
              if (!selected) {
                return;
              }

              props.onChange([...children, selected]);
            }}
            className="add"
            disabled={!selected}
          >
            <AddIcon />
          </Button>
        </Grid>
      </Grid>
    </Grid>
  );
}

export default ChildrenInput;
