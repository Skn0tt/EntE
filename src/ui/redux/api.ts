/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import axios, { AxiosRequestConfig } from "axios";

import {
  APIResponse,
  UserN,
  EntryN,
  SlotN,
  AuthState,
  BasicCredentials,
} from "./types";
import {
  CreateEntryDto,
  CreateUserDto,
  PatchUserDto,
  UserDto,
  EntryDto,
  SlotDto,
  JwtTokenPayload,
  PatchEntryDto,
  Languages,
  InstanceConfigDto,
  LoginDto,
  BlackedUserDto,
  BlackedEntryDto,
  BlackedSlotDto,
} from "@@types";
import * as _ from "lodash";
import { Base64 } from "../helpers/base64";
import { getConfig } from "./config";
import { Map, Set } from "immutable";

const axiosStandardParams = (token: string): AxiosRequestConfig => ({
  ...axiosTokenParams(token),
  validateStatus: (s) => s >= 200 && s < 300,
});

const axiosTokenParams = (token: string): AxiosRequestConfig => ({
  headers: {
    Authorization: "Bearer " + token,
    "X-Authorization": "Bearer " + token
  },
});

const get = async <T>(url: string, token: string) => {
  const response = await axios.get<T>(url, axiosStandardParams(token));
  return response.data;
};

export const mergeAPIResponses = (...values: APIResponse[]): APIResponse => {
  return values.reduce((previousResponse, value) => {
    return {
      entries: previousResponse.entries.mergeDeep(value.entries),
      slots: previousResponse.slots.mergeDeep(value.slots),
      users: previousResponse.users.mergeDeep(value.users),
    };
  }, emptyAPIResponse());
};

const emptyAPIResponse = (): APIResponse => ({
  entries: Map(),
  slots: Map(),
  users: Map(),
});

export const normalizeUsers = (...users: BlackedUserDto[]): APIResponse => {
  let result = emptyAPIResponse();

  users.forEach((user) => {
    const { children = [] } = user;
    const childrenIds = children.map((c) => c.id);

    const normalised = new UserN({
      childrenIds,
      ...user,
      children: [],
    });
    result.users = result.users.mergeIn([normalised.get("id")], normalised);

    const childrenResponse = normalizeUsers(...children);
    result = mergeAPIResponses(result, childrenResponse);
  });

  return result;
};

export const normalizeEntries = (
  ...entries: BlackedEntryDto[]
): APIResponse => {
  let result = emptyAPIResponse();

  entries.forEach((entry) => {
    const { student, slots } = entry;
    const normalised = new EntryN({
      ...entry,
      student: (null as unknown) as UserDto,
      slots: [],
      studentId: student.id,
      slotIds: slots.map((s) => s.id),
    });

    result.entries = result.entries.mergeIn([normalised.get("id")], normalised);

    const slotResponse = normalizeSlots(...slots);
    const studentResponse = normalizeUsers(student);

    result = mergeAPIResponses(result, slotResponse, studentResponse);
  });

  return result;
};

export const normalizeSlots = (...slots: BlackedSlotDto[]): APIResponse => {
  let result = emptyAPIResponse();

  slots.forEach((slot) => {
    const { student, teacher } = slot;
    const normalised = new SlotN({
      ...slot,
      student: (null as unknown) as UserDto,
      teacher: (null as unknown) as UserDto,
      studentId: student.id,
      teacherId: !!teacher ? teacher.id : null,
    });

    result.slots = result.slots.mergeIn([normalised.get("id")], normalised);

    const userResponse =
      teacher === null
        ? normalizeUsers(student)
        : normalizeUsers(student, teacher);
    result = mergeAPIResponses(result, userResponse);
  });

  return result;
};

export const getTokenPayload = (token: string): JwtTokenPayload => {
  const [header, payload, signature] = token.split(".");
  const decodedPayload = JSON.parse(Base64.decode(payload));

  return decodedPayload;
};

export const getAuthState = (token: string): AuthState => {
  const payload = getTokenPayload(token);
  return new AuthState({
    token,
    exp: new Date((payload as any).exp * 1000),
  });
};

export const refreshToken = async (token: string): Promise<AuthState> => {
  try {
    const result = await axios.get<string>(
      "/api/token",
      axiosTokenParams(token)
    );

    return getAuthState(result.data);
  } catch (error) {
    throw error;
  }
};

export const getChildren = async (token: string): Promise<APIResponse> => {
  const data = await get<UserDto[]>(`/api/users?filter=children`, token);
  return normalizeUsers(...data);
};

export const downloadExcelExport = async (token: string): Promise<void> => {
  const response = await axios.get<Blob>(`/api/export/excel`, {
    ...axiosTokenParams(token),
    responseType: "blob",
  });
  getConfig().onFileDownload(response.data, "export.xlsx");
};

export const getEntry = async (
  id: string,
  token: string
): Promise<APIResponse> => {
  const data = await get<EntryDto>(`/api/entries/${id}`, token);
  return normalizeEntries(data);
};

export const getEntries = async (token: string): Promise<APIResponse> => {
  const data = await get<EntryDto[]>(`/api/entries`, token);
  return normalizeEntries(...data);
};

export const getSlots = async (token: string): Promise<APIResponse> => {
  const data = await get<SlotDto[]>(`/api/slots`, token);
  return normalizeSlots(...data);
};

export const getUser = async (
  id: string,
  token: string
): Promise<APIResponse> => {
  const data = await get<UserDto>(`/api/users/${id}`, token);
  return normalizeUsers(data);
};

const _delete = async <T>(url: string, token: string) => {
  const response = await axios.delete(url, axiosTokenParams(token));
  return response.data as T;
};

export const deleteUser = async (id: string, token: string) => {
  await _delete(`/api/users/${id}`, token);
};

export const deleteEntry = async (id: string, token: string) => {
  await _delete(`/api/entries/${id}`, token);
};

export const getUsers = async (token: string): Promise<APIResponse> => {
  const data = await get<UserDto[]>(`/api/users`, token);
  return normalizeUsers(...data);
};

export const post = async <T>(
  url: string,
  token: string,
  body?: {},
  config: AxiosRequestConfig = {}
) => {
  const response = await axios.post<T>(
    url,
    body,
    _.merge(axiosStandardParams(token), config)
  );
  return response.data;
};

export const createEntry = async (
  entry: CreateEntryDto,
  token: string
): Promise<APIResponse> => {
  const response = await post<EntryDto>(`/api/entries/`, token, entry);
  return normalizeEntries(response);
};

export const createUsers = async (
  users: CreateUserDto[],
  token: string
): Promise<APIResponse> => {
  const response = await post<UserDto[]>(`/api/users/`, token, users);
  return normalizeUsers(...response);
};

const patch = async <T, B = {}>(url: string, token: string, body?: B) => {
  const response = await axios.patch<T>(url, body, axiosStandardParams(token));
  return response.data;
};

export const updateUser = async (
  userId: string,
  user: PatchUserDto,
  token: string
): Promise<APIResponse> => {
  const response = await patch<UserDto>(`/api/users/${userId}`, token, user);
  return normalizeUsers(response);
};

const put = async <T, B = {}>(
  url: string,
  token: string,
  body?: B,
  config: AxiosRequestConfig = {}
) => {
  const response = await axios.put<T>(
    url,
    body,
    _.merge(axiosStandardParams(token), config)
  );
  return response.data;
};

export const signEntry = async (
  id: string,
  token: string
): Promise<APIResponse> => {
  const response = await patch<EntryDto, PatchEntryDto>(
    `/api/entries/${id}`,
    token,
    {
      signed: true,
    }
  );
  return normalizeEntries(response);
};

export const setLanguage = async (
  id: string,
  language: Languages,
  token: string
): Promise<void> => {
  await put<void, Languages>(`/api/users/${id}/language`, token, language, {
    transformResponse: [],
    headers: {
      "Content-Type": "text/plain",
    },
  });
};

export const unsignEntry = async (
  id: string,
  token: string
): Promise<APIResponse> => {
  const response = await patch<EntryDto, PatchEntryDto>(
    `/api/entries/${id}`,
    token,
    {
      signed: false,
    }
  );
  return normalizeEntries(response);
};

export const importUsers = async (
  token: string,
  dtos: CreateUserDto[],
  {
    deleteEntries,
    deleteUsers,
    deleteStudentsAndParents,
  }: {
    deleteUsers: boolean;
    deleteEntries: boolean;
    deleteStudentsAndParents: boolean;
  }
) => {
  const result = await post<UserDto[]>(
    `/api/instance/import?deleteUsers=${deleteUsers}&deleteEntries=${deleteEntries}&deleteStudentsAndParents=${deleteStudentsAndParents}`,
    token,
    dtos
  );
  return normalizeUsers(...result);
};

export const fetchInstanceConfig = async (): Promise<InstanceConfigDto> => {
  const response = await axios.get<InstanceConfigDto>(`/api/instanceConfig`);
  const { data } = response;
  return data;
};

export const setLoginBanner = async (
  language: Languages,
  text: string | null,
  token: string
): Promise<void> => {
  const response = await put(
    `/api/instanceConfig/loginBanners/${language}`,
    token,
    text,
    {
      transformResponse: [],
      headers: {
        "Content-Type": "text/plain",
      },
    }
  );
};

export const setDefaultLanguage = async (
  language: Languages,
  token: string
): Promise<void> => {
  const response = await put(
    `/api/instanceConfig/defaultLanguage`,
    token,
    language,
    {
      transformResponse: [],
      headers: {
        "Content-Type": "text/plain",
      },
    }
  );
};

export const setParentSignatureNotificationTime = async (
  value: number,
  token: string
): Promise<void> => {
  const response = await put(
    `/api/instanceConfig/parentSignatureTimes/notification`,
    token,
    "" + value,
    {
      transformResponse: [],
      headers: {
        "Content-Type": "text/plain",
      },
    }
  );
};

export const setParentSignatureExpiryTime = async (
  value: number,
  token: string
): Promise<void> => {
  const response = await put(
    `/api/instanceConfig/parentSignatureTimes/expiry`,
    token,
    "" + value,
    {
      transformResponse: [],
      headers: {
        "Content-Type": "text/plain",
      },
    }
  );
};

export const setEntryCreationDays = async (
  value: number,
  token: string
): Promise<void> => {
  const response = await put(
    `/api/instanceConfig/entryCreationDeadline`,
    token,
    "" + value,
    {
      transformResponse: [],
      headers: {
        "Content-Type": "text/plain",
      },
    }
  );
};

export async function addReviewedRecord(token: string, entryId: string) {
  await axios.post(`/api/reviewedRecords`, entryId, {
    headers: {
      ...axiosStandardParams(token).headers,
      "Content-Type": "text/plain",
    },
  });
}

export async function managerReachdOut(token: string, entryId: string) {
  await post(`/api/entries/${entryId}/managerReachedOut`, token);
}

export const updateManagerNotes = async (
  studentId: string,
  value: string,
  token: string
): Promise<void> => {
  await put(`/api/users/${studentId}/managerNotes`, token, value, {
    transformResponse: [],
    headers: {
      "Content-Type": "text/plain",
    },
  });
};

export async function promoteTeacher(
  teacherId: string,
  _class: string,
  token: string
) {
  await post(`/api/users/${teacherId}/promote`, token, _class, {
    transformResponse: [],
    headers: {
      "Content-Type": "text/plain",
    },
  });
}

export async function demoteManager(managerId: string, token: string) {
  await post(`/api/users/${managerId}/demote`, token);
}
