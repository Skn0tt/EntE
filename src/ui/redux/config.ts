import { Middleware } from "redux";
import { asyncLocalStorage } from "redux-persist/storages";
import { CreateEntryDto, CreateUserDto, PatchUserDto } from "@@types";
import { ImportUsersRequestPayload } from "./actions";

export type SagaListeners = {
  onImportSuccessful: (payload: ImportUsersRequestPayload) => void;
  onUsersCreated: (users: CreateUserDto[]) => void;
  onUserUpdated: (patch: PatchUserDto) => void;
  onSignedEntry: (entryId: string) => void;
  onUnsignedEntry: (entryId: string) => void;
  onEntryCreated: (entry: CreateEntryDto) => void;
  onEntryDeleted: (entryId: string) => void;
  onLoginFailedInvalidCredentials: () => void;
  onUserDeleted: (userId: string) => void;
  onRequestError: (error: Error) => void;
  onSigningError: (error: Error) => void;
  onUnsigningError: (error: Error) => void;
  onSyncingLanguageError: (error: Error) => void;
  onSagaError: (err: Error) => void;
};

export type ReduxConfig = {
  onFileDownload: (file: Blob, filename: string) => void;
  middlewares: Middleware[];
  storage: any;
} & SagaListeners;

const ignore = () => {};

const defaultConfig: ReduxConfig = {
  storage: asyncLocalStorage,
  middlewares: [],
  onFileDownload: ignore,
  onImportSuccessful: ignore,
  onUsersCreated: ignore,
  onEntryCreated: ignore,
  onUserDeleted: ignore,
  onEntryDeleted: ignore,
  onSignedEntry: ignore,
  onRequestError: ignore,
  onSagaError: ignore,
  onSigningError: ignore,
  onSyncingLanguageError: ignore,
  onUnsignedEntry: ignore,
  onUnsigningError: ignore,
  onUserUpdated: ignore,
  onLoginFailedInvalidCredentials: ignore,
};

let config: ReduxConfig = defaultConfig;

export const updateConfig = (v: Partial<ReduxConfig>) => {
  config = { ...config, ...v };
};

export const getConfig = () => config;
