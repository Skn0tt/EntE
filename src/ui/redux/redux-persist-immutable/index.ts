import {
  autoRehydrate as baseAutoRehydrate,
  createPersistor as baseCreatePersistor,
  createTransform,
  getStoredState as baseGetStoredState,
  persistStore as basePersistStore,
  purgeStoredState,
} from "redux-persist";

import immutableTransform from "redux-persist-transform-immutable";
import * as operators from "./operators";
import { stateReconciler } from "./reconciler";
import { Store } from "redux";

const extendConfig = (config: any) => {
  const incomingTransforms = config.transforms || [];
  const records = config.records || null;
  let transforms = incomingTransforms;
  if (!config.skipImmutableTransform) {
    transforms = [...incomingTransforms, immutableTransform({ records })];
  }
  return { stateReconciler, ...config, ...operators, transforms };
};

const autoRehydrate = (config = {}) => {
  return baseAutoRehydrate(extendConfig(config));
};

const createPersistor = (store: Store, config = {}) => {
  return baseCreatePersistor(store, extendConfig(config));
};

const persistStore = (store: Store, config = {}, cb: any) => {
  return basePersistStore(store, extendConfig(config), cb);
};

const getStoredState = (config = {}) => {
  return baseGetStoredState(extendConfig(config));
};

export {
  autoRehydrate,
  createPersistor,
  createTransform,
  getStoredState,
  persistStore,
  purgeStoredState,
};
