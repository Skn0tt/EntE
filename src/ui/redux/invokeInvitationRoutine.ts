import { getConfig } from "./config";
import { post } from "./api";
import { Success, Fail } from "monet";
import { ErrorReporting } from "../ErrorReporting";

const { baseUrl } = getConfig();

export async function invokeInvitationRoutine(userId: string, token: string) {
  try {
    await post(`${baseUrl}/api/users/${userId}/invokeInvitationRoutine`, token);
    return Success<Error, void>(undefined);
  } catch (error) {
    await ErrorReporting.report(error);
    return Fail<Error, void>(error);
  }
}
