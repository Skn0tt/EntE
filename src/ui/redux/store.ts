/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import { createStore as reduxCreateStore, applyMiddleware, Store } from "redux";
import reducer, { initialState } from "./reducer";
import { composeWithDevTools } from "redux-devtools-extension";
import reduxSaga from "redux-saga";
import {
  autoRehydrate,
  persistStore,
  createTransform,
} from "./redux-persist-immutable";
import saga from "./saga";
import {
  AppState,
  IAppState,
  UserN,
  SlotN,
  AuthState,
  EntryN,
  InstanceConfigN,
  ParentSignatureTimesN,
} from "./types";
import * as _ from "lodash";
import { getConfig } from "./config";
import { Persistor } from "redux-persist";

const keysToRemove: (keyof IAppState)[] = ["pendingActions"];
const removeUnneededTransform = createTransform(
  (inboundState: AppState, key: string) =>
    keysToRemove.indexOf(key as keyof IAppState) !== -1
      ? initialState.get(key as keyof IAppState)
      : inboundState,
  _.identity
);

const startPersistance = (store: Store) => {
  return new Promise<Persistor>((resolve) => {
    const persistor = persistStore(
      store,
      {
        storage: getConfig().storage,
        transforms: [removeUnneededTransform],
        records: [
          UserN,
          SlotN,
          AuthState,
          AppState,
          EntryN,
          InstanceConfigN,
          ParentSignatureTimesN,
        ],
      },
      () => {
        resolve(persistor);
      }
    );
  });
};

let store: Store | null = null;
let persistor: Persistor;

export const createStore = async () => {
  const sagaMiddleware = reduxSaga({
    onError: getConfig().onSagaError,
  });
  const composeEnhancers = composeWithDevTools({});

  const middlewares = getConfig().middlewares || [];

  store = reduxCreateStore(
    reducer as any,
    composeEnhancers(
      applyMiddleware(sagaMiddleware, ...middlewares),
      autoRehydrate()
    )
  );

  sagaMiddleware.run(saga);

  persistor = await startPersistance(store);

  return store;
};

export default (store as unknown) as Store;
