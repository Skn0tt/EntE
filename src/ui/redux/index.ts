/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import { createStore } from "./store";
import { updateConfig, ReduxConfig } from "./config";

export * from "./selectors";
export * from "./actions";
export * from "./types";
export * from "./constants";
export * from "./config";

const setup = (conf: Partial<ReduxConfig>) => {
  updateConfig(conf);

  return await createStore();
};

export default setup;
