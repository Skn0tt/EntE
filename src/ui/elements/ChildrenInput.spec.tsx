/**
 * EntE
 * (c) 2017-present, Simon Knott <info@simonknott.de>
 *
 * This source code is licensed under the GNU Affero General Public License
 * found in the LICENSE file in the root directory of this source tree.
 */

import * as React from "react";
import ChildrenInput, { includes } from "./ChildrenInput";
import { shallow } from "enzyme";
import { Roles } from "@@types";
import * as sinon from "sinon";
import { UserN } from "../redux";
import { TestWrapper } from "test/TestWrapper";

const userId = "fdas90ß9sß0";
const user = new UserN({
  id: userId,
  children: [],
  firstName: "Horst",
  lastName: "Hansen",
  displayname: "Horst Hansen",
  email: "hort@hansen.de",
  birthday: "2100-01-01",
  role: Roles.STUDENT,
  username: "hhansen",
});
const user2 = new UserN({
  id: "jkldsfjlk",
  children: [],
  firstName: "Hanni",
  lastName: "Ball",
  displayname: "Hanni Ball",
  email: "hanni@ball.de",
  birthday: "2100-01-01",
  role: Roles.STUDENT,
  username: "hball",
});

describe("ChildrenInput", () => {
  it("renders correctly", () => {
    const onChange = sinon.spy();

    const comp = shallow(
      <TestWrapper>
        <ChildrenInput
          children={[user]}
          onChange={onChange}
          students={[user, user2]}
        />
      </TestWrapper>
    );

    expect(comp).toMatchSnapshot();
  });

  it("renders when no additional users are available", () => {
    const onChange = sinon.spy();
    const comp = shallow(
      <TestWrapper>
        <ChildrenInput
          children={[user]}
          onChange={onChange}
          students={[user]}
        />
      </TestWrapper>
    );

    expect(comp).toMatchSnapshot();
  });
});

describe("includes", () => {
  it("returns false on values from parameter", () => {
    const exc = includes([user]);
    expect(exc(user)).toBe(true);
  });
  it("returns false on values not in parameters", () => {
    const exc = includes([user]);
    expect(exc(user2)).toBe(false);
  });
});
