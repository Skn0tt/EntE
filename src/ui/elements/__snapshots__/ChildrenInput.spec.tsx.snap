// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChildrenInput renders correctly 1`] = `
<WithStyles(Grid)
  container={true}
  direction="column"
>
  <WithStyles(Grid)
    item={true}
  >
    <WithStyles(Typography)
      variant="h6"
    >
      Children
    </WithStyles(Typography)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
  >
    <WithStyles(List)>
      <WithStyles(ListItem)
        key="0"
      >
        <WithStyles(ListItemText)
          primary="Horst Hansen"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(IconButton)
            onClick={[Function]}
          >
            <pure(DeleteIcon) />
          </WithStyles(IconButton)>
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
    </WithStyles(List)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    container={true}
    item={true}
  >
    <WithStyles(Grid)
      item={true}
      xs={11}
    >
      <SearchableDropdown
        helperText="Add children"
        includeItem={[Function]}
        itemToString={[Function]}
        items={
          Array [
            Immutable.UserN {
              "id": "jkldsfjlk",
              "username": "hball",
              "children": Array [],
              "firstName": "Hanni",
              "lastName": "Ball",
              "displayname": "Hanni Ball",
              "email": "hanni@ball.de",
              "birthday": "2100-01-01",
              "role": "student",
              "isAdmin": false,
              "childrenIds": Array [],
              "class": undefined,
              "language": undefined,
              "managerNotes": undefined,
              "subscribedToWeeklySummary": undefined,
            },
          ]
        }
        label="Child"
        onSelect={[Function]}
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={1}
    >
      <WithStyles(Button)
        className="add"
        disabled={false}
        mini={true}
        onClick={[Function]}
        variant="fab"
      >
        <pure(AddIcon) />
      </WithStyles(Button)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;

exports[`ChildrenInput renders when no additional users are available 1`] = `
<WithStyles(Grid)
  container={true}
  direction="column"
>
  <WithStyles(Grid)
    item={true}
  >
    <WithStyles(Typography)
      variant="h6"
    >
      Children
    </WithStyles(Typography)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
  >
    <WithStyles(List)>
      <WithStyles(ListItem)
        key="0"
      >
        <WithStyles(ListItemText)
          primary="Horst Hansen"
        />
        <WithStyles(ListItemSecondaryAction)>
          <WithStyles(IconButton)
            onClick={[Function]}
          >
            <pure(DeleteIcon) />
          </WithStyles(IconButton)>
        </WithStyles(ListItemSecondaryAction)>
      </WithStyles(ListItem)>
    </WithStyles(List)>
  </WithStyles(Grid)>
  <WithStyles(Grid)
    container={true}
    item={true}
  >
    <WithStyles(Grid)
      item={true}
      xs={11}
    >
      <SearchableDropdown
        helperText="Add children"
        includeItem={[Function]}
        itemToString={[Function]}
        items={Array []}
        label="Child"
        onSelect={[Function]}
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={1}
    >
      <WithStyles(Button)
        className="add"
        disabled={true}
        mini={true}
        onClick={[Function]}
        variant="fab"
      >
        <pure(AddIcon) />
      </WithStyles(Button)>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;
